### PR TITLE
Migrate to persistent cluster settings

### PR DESCRIPTION
**Pull request by girishc13**
The transient cluster settings are deprecated since 7.16.0. In order to support the deprecation and future removal, the es-operator will now only edit the persistent cluster settings for scaling operations.

## Description
To ensure a smooth migration, transient settings will be first copied to the persistent settings and then the es-operator will  operate exclusively with persistent settings. The transient settings are first copied over existing persistent settings because the transient settings have the highest priority by default.

## Types of Changes
- extract ESSettings anonymous typed to explicit types
- introduce handy methods for reading cluster settings from ESSettings type
- introduce gregu/null library for easy handling of nil value serialization required for removing  existing transient settings.
- copy over/merge non empty transient settings to persistent settings
- operate on persistent settings only

## Tasks
_List of tasks you will do to complete the PR_
  - [x] Add more unit tests to ensure that overwriting persistent settings from non empty transient settings ensure proper cluster state.
  - [x] Update documentation and add a migration guide to specify the breaking change.

## Review
_List of tasks the reviewer must do to review the PR_
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: operator/es_client.go [hunk 1]
```diff
@@ -67,22 +68,68 @@ type ESHealth struct {
...
+		esSettings.Persistent.Cluster.Routing.Rebalance.Enable = null.StringFromPtr(&value)
+		esSettings.Transient.Cluster.Routing.Rebalance.Enable = null.StringFromPtr(nil)
+	}
+
+	if transientExcludeIps := esSettings.GetTransientExcludeIPs().ValueOrZero(); transientExcludeIps != "" {
+		persistentExcludeIps := esSettings.GetPersistentExcludeIPs().ValueOrZero()
+		if persistentExcludeIps == "" {
+			esSettings.Persistent.Cluster.Routing.Allocation.Exclude.IP = null.StringFromPtr(&transientExcludeIps)
+		} else {
+			mergedIps := null.StringFrom(transientExcludeIps + "," + persistentExcludeIps)
```
Comment (otrosien) [inline operator/es_client.go:112]: do we need to deduplicate?

Comment (girishc13) [reply operator/es_client.go:112]: I don't think it's strictly required for the es-operator manipulations because the new exclude ip list will also have this duplicated entry. It might cause some confusion if for some reason a developer interfered and is manually updating the ip exclusion list and doesn't notice the duplicate entry. It's better to deduplicate to avoid such edge cases.

File: go.mod [hunk 2]
```diff
@@ -9,6 +9,7 @@ require (
 	github.com/stretchr/testify v1.7.0
 	golang.org/x/tools v0.1.1 // indirect
 	gopkg.in/alecthomas/kingpin.v2 v2.2.6
+	gopkg.in/guregu/null.v4 v4.0.0
```
Comment (otrosien) [inline go.mod:12]: if we only use 1-2 methods, we should be copying it over into our codebase

**Comment by otrosien** [comment]
👍

**Comment by otrosien** [comment]
👍

**Comment by girishc13** [comment]
👍

**Comment by otrosien** [comment]
👍
